4.26/platform_isv.html

```diff
@@ -38,6 +38,81 @@ <h2>Platform and Equinox API</h2>
+race condition. The listener could not deterministically know if the job
+was still scheduled or already done. As a consequence a join() could
+have returned too early or the Progress View did not show jobs that did
+still run. And probably many other things went wrong totally unnoticed.
+Even multiple notifications could have happened in parallel in various
+worker threads, as there is no guarantee that the next execution is done
+in the same worker.</p>
+<p>To fix this indeterministic behavior all events for the same Job instance will
+now be sent one after the other. It is still not defined in which thread
+job events will be sent, but the new implementation will not call any
+IJobChangeListener until all pending events for the same job instance are processed
+by all registered IJobChangeListener. The new implementation will also
+wait with any job state change until all calls to IJobChangeListener for
+that job returned.</p>
+<p>The IJobChangeListener javadoc clearly specifies &quot;whether the state
+change occurs before, during, or after listeners are notified is
+unspecified.&quot; - and the implementation changed within that broad
+specification. Unfortunately some IJobChangeListener around rely on the
+old implementation. They may now deadlock. For example the following
+snippet can now deadlock:</p>
+<pre><code> job.schedule(); <span class="hljs-comment">// may result in done();</span>
+ <span class="hljs-keyword">synchronized</span> (lock){
+  <span class="hljs-keyword">boolean</span> schedule = ...; <span class="hljs-comment">// lock needed for reasons</span>
+  <span class="hljs-keyword">if</span> (schedule ) job.schedule(); <span class="hljs-comment">// schedule() will notify scheduled()</span>
+  <span class="hljs-comment">// - which may not return before previous Notifications return!</span>
+ }
+
+ <span class="hljs-keyword">public</span> <span class="hljs-function"><span class="hljs-keyword">void</span> <span class="hljs-title">done</span><span class="hljs-params">(IJobChangeEvent event)</span> </span>{
+  <span class="hljs-comment">// can not enter while other Thread holds lock:</span>
+  <span class="hljs-keyword">synchronized</span> (lock) {<span class="hljs-comment">// possible deadlock</span>
+  }
+ }
+</code></pre><p>Instead use:</p>
+<pre><code> job.schedule();
+ <span class="hljs-keyword">boolean</span> schedule;
+ <span class="hljs-keyword">synchronized</span> (lock) {
+  schedule=...; <span class="hljs-comment">// lock needed for reasons</span>
+ }
+ <span class="hljs-keyword">if</span> (schedule) job.schedule();
+
+ <span class="hljs-keyword">public</span> <span class="hljs-function"><span class="hljs-keyword">void</span> <span class="hljs-title">done</span><span class="hljs-params">(IJobChangeEvent event)</span> </span>{
+  <span class="hljs-keyword">synchronized</span> (lock) {<span class="hljs-comment">// OK</span>
+  }
+ }
+</code></pre><p>The same problem may occur on all IJobChangeListener methods and all
```

Comment:
One probably could also add "bad" example from https://github.com/eclipse-platform/eclipse.platform/pull/262 and the improved version, because it looks "not wrong" and works in 99% of cases as we saw, but not in 100%...

```
Object syncObject = new Object();
job1.addJobChangeListener(new JobChangeAdapter() {
	@Override
	public void running(IJobChangeEvent event) {
		job2.schedule();
		synchronized (syncObject) {
			syncObject.notify();
		}
	}
});

synchronized (syncObject) {
	job1.schedule();
	syncObject.wait();
}
```

Reply:
The given example already shows that "synchronized" around schedule() is wrong.